### DISCUSSION TITLE
Comprehensive Color Image Normalization / Color Invariance?

### DISCUSSION THREAD

**Discussion by oisact** [ImageMagick/ImageMagick#5839] [Help]
I've scoured the documentation and don't see where ImageMagick can do Comprehensive Color Image Normalization. Just wondering it perhaps it is called something different, or if that can be achieved through a series of already-available routines?

Or another option is color invariance processing. As described in various publications, such as:

http://www2.cmp.uea.ac.uk/Research/groups/compvis/ColourInvariants/ColourInvariants.htm#FIN.XU.CIC10

This is generally used for AI identification of objects, as it is essentially reducing an image to the pure colors regardless of how it is illuminated.

My goal here is to perform this kind of normalization to a pair of images, which will then be compared with one another. This will help prevent lighting changes (shadows, etc) from showing up as differences in the comparison.

**Comment by fmw42** [comment]
There is no current function in Imagemagick that will do all that. But you can compute the means of the RGB values of each pixel from -intensity average (that is equal to (R+G+B)/3) for each pixel, then get the average over the whole image from -format "%[mean]".  If you want the mean of all R values, then separate the R channel and do -format "%[mean]".

It is not clear to me from that link what they mean by iterative.  Do you just do one then the other once or do you have to repeat both until you get some convergence?

An Imagemagick function that might help is -color-matrix.  See https://imagemagick.org/script/command-line-options.php#color-matrix.

**Comment by snibgo** [comment]
The paper https://www.researchgate.net/publication/2576969_Comprehensive_Colour_Image_Normalization  "Comprehensive Colour Image Normalization", Finlayson et al, 1999, describes an iterative process. Each iteration is: normalize each pixel by (R+B+G) at that pixel, then normalize each channel of each pixel by the overall mean value of that channel. Iterate until the output is stable.

IM contains low-level operations that we can combine to perform a single iteration. It doesn't contain flow-control structures for iteration to stability, so that has to be done in a shell script.

From a quick skim-read of the paper, one iteration can be (Windows BAT syntax):
```
magick ^
  %INFILE% ^
  -alpha off ^
  -define compose:clamp=off ^
  -colorspace RGB ^
  ( +clone ^
    -channel RGB -separate +channel ^
    -evaluate-sequence Add ^
  ) ^
  -compose DivideSrc -composite ^
  ( +clone ^
    -set option:MYSIZE %%wx%%h ^
    -scale "1x1^!" ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -compose DivideSrc -composite ^
  -evaluate Divide 3 ^
  -colorspace sRGB ^
  -define quantum:format=floating-point -depth 32 ^
  %OUTFILE%
```

**Comment by fmw42** [comment]
What OS and what version of Imagemagick is this needed for?

**Reply by oisact** [reply]
I'm experimenting with Windows. I downloaded the primary Windows distribution. `ImageMagick-7.1.0-52-Q16-HDRI-x64-dll.exe`

**Comment by fmw42** [comment]
Sorry, I could script it for Linux/Max OSX/Windows 10 Unix/Cygwin. But not for DOS Windows.

**Comment by snibgo** [comment]
Here is a Windows script I have named `CompColNorm.bat`. It is not thoroughly tested.

```
rem Comprehensive Colour Normalisation, as described in
rem http://www2.cmp.uea.ac.uk/Research/compvis/Papers/FinXu_CGIV02.pdf
rem "Non-iterative Comprehensive Normalisation", Finlayson et al, 1999.
rem
rem %1 input image
rem %2 output image
rem %3 max RMSE at convergence

set INFILE=%1
set OUTFILE=%2
set MaxRmse=%3

if "%MaxRmse%"=="." set MaxRmse=
if "%MaxRmse%"=="" set MaxRmse=1e-4


set TMPFILE=%TEMP%\ccn.miff


%IMG7%magick ^
  %INFILE% ^
  -define quantum:format=floating-point -depth 32 ^
  %TMPFILE%

:loop

for /F "usebackq" %%L in (`%IMG7%magick ^
  %TMPFILE% ^
  ^( +clone ^
    -alpha off ^
    -define "compose:clamp=off" ^
    -colorspace RGB ^
    ^( +clone ^
      -channel RGB -separate +channel ^
      -evaluate-sequence Add ^
    ^) ^
    -compose DivideSrc -composite ^
    ^( +clone ^
      -set option:MYSIZE %%wx%%h ^
      -scale "1x1^!" ^
      -scale "%%[MYSIZE]^!" ^
    ^) ^
    -compose DivideSrc -composite ^
    -evaluate Divide 3 ^
    -colorspace sRGB ^
    -define "quantum:format=floating-point" -depth 32 ^
    -write %TMPFILE% ^
  ^) ^
  -metric RMSE ^
  -format "diff=%%[distortion]\nDoAgain=%%[fx:%%[distortion]>%MaxRmse%]\n" ^
  -compare ^
  info:`) do set %%L

if %DoAgain%==1 goto loop

%IMG7%magick ^
  %TMPFILE% ^
  %OUTFILE%
```
